### ISSUE TITLE
shared-only fields

### ISSUE THREAD

**Issue by nikomatsakis** [nikomatsakis/fields-in-traits-rfc#5]
The current RFC proposes only "mutable" fields. That is, if you have an `&mut T`, and `T: Trait`, you can mutate the fields declared in `Trait`. However, it'd be nice to be able to have "shared-only" fields, where the fields act shared no matter what. So even with a `&mut T`, you can only get a shared borrow of those fields.

Among other things, this is needed if we ever wish to support defining traits with fields that can be implemented by `&T` types. Example:

```rust
trait Foo {
    value: u32
}

impl Foo for &u32 {
    value = *self
}
```

Right now, though, that is blocked by #4 in any case.

**Comment by glaebhoerl** [comment]
(What does it mean when a `P-essential` is blocked on a `P-nice-to-have`?)

Anyway, this seems like a nice-to-have to me, but awkardly, syntax-wise the most obvious (and likely overall best) way to signify this would be by the *absence* of a `mut` qualifier - that is, "normal" mutable fields would be `let mut` or whatever. So we probably want to decide one way or another on the syntax aspect at least before *stabilizing* the core feature.

(Edit: I see you mentioned the same thing in #3.)

**Comment by burdges** [comment]
I think "blocked" means #4 requires adding `mut` here so that smart pointer `Deref`s can do their thing.

**Comment by nikomatsakis** [comment]
@glaebhoerl 

> (What does it mean when a P-essential is blocked on a P-nice-to-have?)

Are you referring to #4? This particular issue is not blocked on #4, I don't think, it just "interacts" with it. (In particular, combining the two enables fields that are reachable through an `&`-ref.)
